src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs

```diff
@@ -0,0 +1,655 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections.Internal;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities;
+
+#pragma warning disable CA1068 // CancellationToken parameters must come last
+#pragma warning disable IDE0007 // Use implicit type
+#pragma warning disable IDE2003 // Blank line required between block and subsequent statement
+#pragma warning disable IDE2004 // Blank line not allowed after constructor initializer colon
+#pragma warning disable VSTHRD200 // Use "Async" suffix for async methods
+
+#if !NET
+
+// Ported from
+// https://github.com/dotnet/runtime/blob/main/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs
+// With only changes to make the code work on NetFx.  Where changes have been made, the original code is kept around in
+// an ifdef'ed block to see what it was doing.
+
+internal static partial class RoslynParallel
+{
+    private static class NetFramework
+    {
+        /// <summary>Executes a for each operation on an <see cref="IEnumerable{TSource}"/> in which iterations may run in parallel.</summary>
+        /// <typeparam name="TSource">The type of the data in the source.</typeparam>
+        /// <param name="source">An enumerable data source.</param>
+        /// <param name="body">An asynchronous delegate that is invoked once per element in the data source.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="source"/> argument or <paramref name="body"/> argument is <see langword="null"/>.</exception>
+        /// <returns>A task that represents the entire for each operation.</returns>
+        /// <remarks>The operation will execute at most <see cref="Environment.ProcessorCount"/> operations in parallel.</remarks>
+        public static Task ForEachAsync<TSource>(IEnumerable<TSource> source, Func<TSource, CancellationToken, ValueTask> body)
+        {
+#if false
+            ArgumentNullException.ThrowIfNull(source);
+            ArgumentNullException.ThrowIfNull(body);
+#endif
```

Comment:
for now, just disabling the arg checks internally.  we rarely ever check this with our ienumerable helpers.